// msw/lib/core/experimental/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var experimental_exports = {};
__export(experimental_exports, {
  HandlersController: () => import_handlers_controller.HandlersController,
  HttpNetworkFrame: () => import_http_frame.HttpNetworkFrame,
  InMemoryHandlersController: () => import_handlers_controller.InMemoryHandlersController,
  InterceptorSource: () => import_interceptor_source.InterceptorSource,
  NetworkFrame: () => import_network_frame.NetworkFrame,
  NetworkSource: () => import_network_source.NetworkSource,
  WebSocketNetworkFrame: () => import_websocket_frame.WebSocketNetworkFrame,
  defineNetwork: () => import_define_network.defineNetwork
});
module.exports = __toCommonJS(experimental_exports);
var import_define_network = require("./define-network");
var import_network_source = require("./sources/network-source");
var import_interceptor_source = require("./sources/interceptor-source");
var import_network_frame = require("./frames/network-frame");
var import_http_frame = require("./frames/http-frame");
var import_websocket_frame = require("./frames/websocket-frame");
var import_handlers_controller = require("./handlers-controller");
